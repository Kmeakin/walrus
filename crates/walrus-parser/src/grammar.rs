use nom::Parser;
use nom_supreme::parser_ext::ParserExt;
use walrus_lexer::Token;
use walrus_syntax::{nodes::*, tokens::*};

mod support;
mod tokens;

mod decl;
mod expr;
mod lit;
mod pat;
mod ty;

pub use self::{decl::*, expr::*, lit::*, pat::*, support::*, tokens::*, ty::*};

#[cfg(test)]
use crate::test_parse;

pub type Err = ();
pub type IResult<'i, T> = nom::IResult<Input<'i>, T, Err>;
pub type Input<'i> = &'i [Token<'i>];

#[cfg(test)]
#[macro_export]
macro_rules! test_parse {
    ($name:ident, $parser:expr, $src:expr) => {
        #[test]
        fn $name() { test_parse($parser, $src) }
    };
}

#[cfg(test)]
fn test_parse<T>(mut parser: impl Fn(Input) -> IResult<T>, src: &'static str)
where
    T: std::fmt::Debug,
{
    use insta::*;

    let tokens = walrus_lexer::lex(src)
        .filter(|token| !token.kind.is_trivia())
        .collect::<Vec<_>>()
        .clone();
    let got = parser.parse(&tokens);

    let mut settings = insta::Settings::new();
    settings.set_snapshot_path("../snapshots");
    settings.set_prepend_module_to_snapshot(false);
    settings.bind(|| assert_debug_snapshot!(got));
}
