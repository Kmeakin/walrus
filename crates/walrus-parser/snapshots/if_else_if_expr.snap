---
source: crates/walrus-parser/src/grammar.rs
expression: got
---
Ok(
    (
        [],
        If(
            IfExpr {
                kw_if: KwIf {
                    span: 0..2,
                },
                test_expr: Lit(
                    Bool(
                        True(
                            KwTrue {
                                span: 3..7,
                            },
                        ),
                    ),
                ),
                then_block: Block {
                    lcurly: LCurly {
                        span: 8..9,
                    },
                    stmts: [],
                    expr: None,
                    rcurly: RCurly {
                        span: 9..10,
                    },
                },
                else_expr: Some(
                    ElseIf {
                        kw_else: KwElse {
                            span: 11..15,
                        },
                        if_expr: IfExpr {
                            kw_if: KwIf {
                                span: 16..18,
                            },
                            test_expr: Lit(
                                Bool(
                                    False(
                                        KwFalse {
                                            span: 19..24,
                                        },
                                    ),
                                ),
                            ),
                            then_block: Block {
                                lcurly: LCurly {
                                    span: 25..26,
                                },
                                stmts: [],
                                expr: None,
                                rcurly: RCurly {
                                    span: 26..27,
                                },
                            },
                            else_expr: None,
                        },
                    },
                ),
            },
        ),
    ),
)
